### PR TITLE
Route handling & Storage

### PR DESCRIPTION
**Pull request by joanromano**

### CODE REVIEW

File: Source/KakapoServer.swift [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+//
+//  Created by Joan Romano on 31/03/16.
+//  Copyright © 2016 devlucky. All rights reserved.
+//
+
+import Foundation
+
+class KakapoServer: NSURLProtocol {
+    
+    typealias KakapoLookupObject = (method: KakapoServerMethod,
```
Comment (zzarcon) [inline Source/KakapoServer.swift]: `routes`

File: Source/KakapoServer.swift [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+                                    handler: (request: KakapoRequest) -> ())
+    
+    enum KakapoServerMethod: String {
```
Comment (zzarcon) [inline Source/KakapoServer.swift]: `KakapoHTTPMethod`

**Comment by MP0w** [comment]
We are actually prefixing stuff (`Kakapo`) without reasons

File: Source/NSURL+Kakapo.swift [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+//  NSURL+Kakapo.swift
+//  KakapoExample
+//
+//  Created by Joan Romano on 31/03/16.
+//  Copyright © 2016 devlucky. All rights reserved.
+//
+
+import Foundation
+
+func checkUrl(handlerUrl: String, requestUrl: String) -> [String : String]? {
```
Comment (zzarcon) [inline Source/NSURL+Kakapo.swift]: What about `getParams` or `paramsFor`?

File: Source/KakapoServer.swift [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    override func startLoading() {
+        guard let requestString = request.URL?.absoluteString else { return }
+        
+        for (key, object) in KakapoServer.routes {
+            if let params = checkUrl(key, requestUrl: requestString) {
+                object.handler(request: KakapoRequest(urlString: requestString, params: params))
+            }
+        }
+        
+        let response = NSHTTPURLResponse(URL: request.URL!, statusCode: 200, HTTPVersion: "", headerFields: nil)
```
Comment (zzarcon) [inline Source/KakapoServer.swift]: We must provide a flexible way of return `statusCode`. I'll suggest to let the user to return the status code from the `requestHandler`, since is the place in which we have all the context, think about the common case of making a request for an unexisting user or asking for a user which doesn't have permission to that enpoint. Is in the request handler where we know about that...

So I'll like to be able to do something like

``` swift
let statusCode = object.handler(request: KakapoRequest(urlString: requestString, params: params))

let response = NSHTTPURLResponse(URL: request.URL!, statusCode: statusCode, HTTPVersion: "", headerFields: nil)

```

 @MP0w @joanromano

**Comment by zzarcon** [comment]
@joanromano we should add tests for `checkUrl` method :crossed_flags:

**Comment by zzarcon** [comment]
Actually, I think we should rename `checkUrl` into `parseUrl`

**Comment by joanromano** [comment]
@MP0w this is ready to be merged

File: Source/NSURL+Kakapo.swift [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+//
+//  Created by Joan Romano on 31/03/16.
+//  Copyright © 2016 devlucky. All rights reserved.
+//
+
+import Foundation
+
+typealias URLInfo = (params: [String : String], queryParams: [String : String])
+
+/// parseUrl: Checks and parses if a given `handleURL` representation matches a `requestURL`. Examples:
```
Comment (MP0w) [inline Source/NSURL+Kakapo.swift:13]: @joanromano weird documentation style, are you using VVDocumenter?

File: Source/KakapoServer.swift [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+    
+    struct KakapoRequest {
+        let urlString: String
+        let info: URLInfo
+    }
+    
+    private static var routes: [String : KakapoLookupObject] = [:]
+    
+    class func enable() {
+        NSURLProtocol.registerClass(self)
```
Comment (MP0w) [inline Source/KakapoServer.swift:28]: what happens if enabled is called twice? Is it possible?

File: Source/NSURL+Kakapo.swift (hunk 5 above)
Comment (joanromano) [inline Source/NSURL+Kakapo.swift:13]: Using this http://ericasadun.com/2015/06/14/swift-header-documentation-in-xcode-7/

File: Source/KakapoServer.swift (hunk 6 above)
Comment (joanromano) [inline Source/KakapoServer.swift:28]: I guess it should have no effects? Not sure if we should handle that failure case in NSURLProtocol (if any) but could be an option.

File: Source/NSURL+Kakapo.swift [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+///
+///
+///     `/users/:id` with `/users/1` produces `queryParams: ["id" : "1"]`
+///     `/users/:id/comments` with `/users/1/comments` produces `queryParams: ["id" : "1"]`
+///     `/users/:id/comments/:comment_id` with `/users/1/comments/2?page=2&author=hector` produces `queryParams: ["id": "1", "comment_id": "2"]` and `params: ["page": "2", "author": "hector"]`
+///
+/// - Parameter handlerURL: the URL with dynamic paths
+/// - Parameter requestURL: the actual URL
+/// - Returns: URLParams
+func parseUrl(handlerURL: String, requestURL: String) -> URLInfo? {
```
Comment (MP0w) [inline Source/NSURL+Kakapo.swift:23]: cool @joanromano but as you said we may use NSURL which already supports all http://nshipster.com/nsurl/ otherwise we are reinventing the wheel and probably not unit testing enough ?

File: Source/String+Kakapo.swift [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+//  KakapoExample
+//
+//  Created by Joan Romano on 31/03/16.
+//  Copyright © 2016 devlucky. All rights reserved.
+//
+
+import Foundation
+
+func randomString(withLength len: Int = 20) -> String {
+    let allowedChars = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
```
Comment (MP0w) [inline Source/String+Kakapo.swift:12]: why is this here if is used only for unit test? why not using `NSUUID` ?

File: Source/NSURL+Kakapo.swift (hunk 7 above)
Comment (MP0w) [inline Source/NSURL+Kakapo.swift:23]: oh is not doable for the dynamic URL

File: Source/KakapoDB.swift [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+//  KakapoExample
+//
+//  Created by Joan Romano on 31/03/16.
+//  Copyright © 2016 devlucky. All rights reserved.
+//
+
+import Foundation
+
+protocol KStorable {
+    var id: Int { get }
```
Comment (MP0w) [inline Source/KakapoDB.swift:12]: id might be a string I guess and use `NSUUID`, no? otherwise we need a thread safe uuid (KakapoDB is not) or random one

File: Source/KakapoDB.swift [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+    init(id: Int)
+}
+
+enum KakapoDBError: ErrorType {
+    case InvalidId
+}
+
+class KakapoDB {
+    
+    private var _uuid = 0
```
Comment (MP0w) [inline Source/KakapoDB.swift:22]: not thread safe

File: Source/NSURL+Kakapo.swift (hunk 7 above)
Comment (joanromano) [inline Source/NSURL+Kakapo.swift:23]: We could use NSURL for the URLComponents splitting, but not sure if it's worth it for all the other stuff.

File: Source/KakapoDB.swift (hunk 9 above)
Comment (joanromano) [inline Source/KakapoDB.swift:12]: Why not an Int? Also, since we are clearly _not_ building a database, do we really need to care about thread safety regarding that id? Are we accessing this from multiple threads in our unit tests environments?

File: Source/String+Kakapo.swift (hunk 8 above)
Comment (joanromano) [inline Source/String+Kakapo.swift:12]: Might be used when creating serializable structs, as the examples in the unit tests.

File: Source/KakapoDB.swift (hunk 9 above)
Comment (MP0w) [inline Source/KakapoDB.swift:12]: I think yes since you are asserting if the id is wrong, also how do you deal with it with the scenarios?
Thread safety for the id (if we keep it incremental) is a must have for me, I guess it could be common to have threading within the app that shouldn't be fUcked by our lib
